demo/site/src/middleware.ts

```diff
@@ -1,36 +1,9 @@
-import type { NextRequest } from "next/server";
-import { NextResponse } from "next/server";
+import { withAdminRedirectMiddleware } from "./middleware/adminRedirect";
+import { chain } from "./middleware/chain";
+import { withDamRewriteMiddleware } from "./middleware/damRewrite";
+import { withPredefinedPagesMiddleware } from "./middleware/predefinedPages";
 
-import { domain } from "./config";
-import { getPredefinedPageRedirect, getPredefinedPageRewrite } from "./middleware/predefinedPages";
-
-export async function middleware(request: NextRequest) {
-    const { pathname } = new URL(request.url);
-
-    const scope = { domain };
-
-    const predefinedPageRedirect = await getPredefinedPageRedirect(scope, pathname);
-
-    if (predefinedPageRedirect) {
-        return NextResponse.redirect(new URL(predefinedPageRedirect, request.url), 307);
-    }
-
-    const predefinedPageRewrite = await getPredefinedPageRewrite(scope, pathname);
-
-    if (predefinedPageRewrite) {
-        return NextResponse.rewrite(new URL(predefinedPageRewrite, request.url));
-    }
-
-    if (pathname.startsWith("/dam/")) {
-        return NextResponse.rewrite(new URL(`${process.env.API_URL_INTERNAL}${request.nextUrl.pathname}`));
-    }
-
-    if (request.nextUrl.pathname === "/admin" && process.env.ADMIN_URL) {
-        return NextResponse.redirect(new URL(process.env.ADMIN_URL));
-    }
-
-    return NextResponse.next();
-}
+export default chain([withPredefinedPagesMiddleware, withAdminRedirectMiddleware, withDamRewriteMiddleware]);
```

Comment:
Alternatively, we could use a behavior similar to the Next middleware where returning `undefined is the same as `NextResponse.next()`:

```tsx
export async function middleware(request: NextRequest, response: NextResponse) {
    return await withPredefinedPagesMiddleware(request, response) ?? 
        await withAdminRedirectMiddleware(request, response) ??
        await withDamRewriteMiddleware(request, response)
}
```

The separate "middlewares" could then be simplified to:

```tsx
async withAdminRedirectMiddleware(request: NextRequest, response: NextResponse) => {
    if (request.nextUrl.pathname === "/admin" && process.env.ADMIN_URL) {
        return NextResponse.redirect(new URL(process.env.ADMIN_URL));
    }
}
```

What do you think?